### PR TITLE
feature(data): add types to dispatch and select

### PR DESCRIPTION
**Pull request by johnhooks**
## What?

Add TypeScript types to `dispatch` and `select` of `@wordpress/data`.

## Why?

I and others are using these function with custom stores and would like to use them fully typed. I am currently [hacking](https://github.com/WordPress/wp-feature-notifications/blob/edc66d533bc873f81d9cd5ae7a9e3e871e1c6301/types/global.d.ts#L9-L12) the type system to accomplish this in the [`wp-feature-notifications`](https://github.com/WordPress/wp-feature-notifications) project, but it would be better if these were typed correctly.

## How?

I moved the functions to individual TypeScript files because I couldn't provide the correct generic arguments in the `index.js` file like in [`useSelect`](https://github.com/WordPress/gutenberg/blob/0b5e6129bc76c41ec252fae5e17277922fe48f12/packages/data/src/components/use-select/index.js#L22-L36) because the types would be exports from packages with the generic arguments. Moving them to a separate file solved this issue, I made them individual files because it seemed appropriate. 

@swissspidy @adamziel @dmsnell 

## Issues

The automatically built docs don't appropriately show the generic type arguments.

### CODE REVIEW

**Comment by swissspidy** [comment]
Awesome, thank you! I‘ll give this a try as soon as I am back from vacation.

I think after this, thunks are the remaining main thing still without proper types.

**Comment by swissspidy** [comment]
This seems to work well in my project 👍

**Review by swissspidy — APPROVED** [review]

**Comment by johnhooks** [comment]
@swissspidy that's awesome. This is my very first PR to Gutenberg that has had an "approved" review! I'm really new to this, what is the next step?

**Comment by swissspidy** [comment]
Normally I would just merge the PR, but since I am not super familiar with the `data` package and its types, it would be great to wait for another review, ideally by a code owner.

**Review by adamziel — APPROVED** [review]
Looks good, thank you!

**Comment by johnhooks** [comment]
I'm so pumped, my first ever PR merged into Gutenberg 🎉

**Pull request merged**
